### ISSUE TITLE
Interpreter discovery is run multiple times for one host if a module is used in a loop which returns ansible_facts

### ISSUE THREAD

**Issue by felixfontein** [ansible/ansible#56234]
##### SUMMARY
If a module which returns `ansible_facts` (just `{}` will do) is run in a loop, interpreter discovery is ran more than once for the same host.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
core

##### ANSIBLE VERSION
```
2.8.0rc2
2.9.0.dev0
```

##### STEPS TO REPRODUCE
```
---
- hosts: localhost
  tasks:
    - setup:
      loop:
        - a
        - b
```
If this is `test.yml`, run `ansible-playbook -vvv test.yml | grep "Attempting python interpreter discovery"`

##### EXPECTED RESULTS
```
<localhost> Attempting python interpreter discovery
```

##### ACTUAL RESULTS
```
<localhost> Attempting python interpreter discovery
<localhost> Attempting python interpreter discovery
```

**Comment by felixfontein** [comment]
Even works with a trivial module:
```.py
from ansible.module_utils.basic import AnsibleModule
if __name__ == '__main__':
    mod = AnsibleModule(argument_spec=dict())
    mod.exit_json(ansible_facts=dict())
```

**Comment by felixfontein** [comment]
At this point `ansible_facts.discovered_interpreter_python` available from the previous loop iteration is removed, since `namespace_facts(af)` is of the form `{"ansible_facts": {...}}`: https://github.com/ansible/ansible/blob/devel/lib/ansible/executor/task_executor.py#L759

**Comment by apollo13** [comment]
I am getting similar warnings when delegating setup from a centos 7 to an ubuntu 18.04 host:
```
- name: Regather facts to get SSH host keys
  setup:
    gather_subset: "!all"
  delegate_to: "{{ item }}"
  delegate_facts: yes
  loop: ...
```
results in:
```
    TASK [borgbackup : Regather facts to get SSH host keys] ************************
    ok: [borg-centos-system -> borg-srv-centos] => (item=borg-srv-centos)
    ok: [borg-centos-system -> borg-srv-ubuntu] => (item=borg-srv-ubuntu)
[DEPRECATION WARNING]: Distribution Ubuntu 18.04 on host borg-centos-system 
should use /usr/bin/python3, but is using /usr/bin/python for backward 
compatibility with prior Ansible releases. A future Ansible release will 
default to using the discovered platform python for this host. See https://docs
.ansible.com/ansible/2.8/reference_appendices/interpreter_discovery.html for 
more information. This feature will be removed in version 2.12. Deprecation 
warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
```

The check now thinks that borg-centos-system is an ubuntu system :)

**Comment by bcoca** [comment]
tested with lastest 2.8/2.9/2.10 and i cannot reproduce
``` ansible-playbook -vvv test.yml -i localhost, -c local 2>&1| grep "Python interpreter discovery fallback"
<localhost> Python interpreter discovery fallback (unsupported Linux distribution: gentoo)```

fyi, i had to define localhost, if using implicit it always has `ansible_python_interpreter` defined so discovery was never triggered.

**Comment by felixfontein** [comment]
@bcoca I don't see it either with latest stable-2.9 and stable-2.10, so I guess it has been fixed somewhen... Thanks for checking!

**Comment by bcoca** [comment]
probably fixed indirectly via #71096

**Comment by felixfontein** [comment]
@bcoca yeah, that looks like it should have fixed it.
